File: 16-take-screenshot-on-failure/java/src/test/java/Screenshot.java
package src.test.java;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.junit.rules.TestWatcher;
import org.junit.runner.Description;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.TakesScreenshot;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import java.io.File;
import java.io.IOException;
import org.apache.commons.io.FileUtils;
import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.MatcherAssert.assertThat;

public class Screenshot {
    WebDriver driver;

    @Before
    public void setUp() throws Exception {
        driver = new FirefoxDriver();
    }

    @Rule
    public TestRule watcher = new TestWatcher() {
        @Override
        protected void failed(Throwable throwable, Description description) {
            File scrFile = ((TakesScreenshot)driver).getScreenshotAs(OutputType.FILE);
            try {
                FileUtils.copyFile(scrFile,
                        new File("failshot_"
                                + description.getClassName()
                                + "_" + description.getMethodName()
                                + ".png"));
            } catch (IOException exception) {
                exception.printStackTrace();
            }
        }

        @Override
        protected void finished(Description description) {
            driver.quit();
        }
    };

    @Test
    public void OnError() {
        driver.get("http://the-internet.herokuapp.com");
        assertThat(false, is(true));
    }

}

File: 16-take-screenshot-on-failure/kotlin/src/test/kotlin/Screenshot.kt
import org.junit.Rule
import org.junit.rules.TestWatcher
import org.junit.runner.Description
import org.openqa.selenium.firefox.FirefoxDriver
import java.io.File
import kotlin.test.assertFalse
import org.junit.Test
import ru.yandex.qatools.ashot.AShot
import ru.yandex.qatools.ashot.shooting.ShootingStrategies
import javax.imageio.ImageIO

class Screenshot {

    private val driver by lazy {
        FirefoxDriver()
    }

    @get:Rule
    val watcher: TestWatcher = object : TestWatcher() {
        override fun failed(e: Throwable?, description: Description?) {
            helperAShot("fail")
        }

        override fun succeeded(description: Description?) {
            helperAShot("success")
        }

        override fun finished(description: Description?) {
            // Quit session (closes browser)
            driver.quit()
        }
    }

    private fun helperAShot(fileName: String) {
        val screenshot = AShot().shootingStrategy(ShootingStrategies.viewportPasting(1000)).takeScreenshot(driver)
        ImageIO.write(screenshot.image, "PNG", File("~/screenshots/$fileName.png"))
    }

    @Test
    fun onError() {
        driver.get("http://the-internet.herokuapp.com")
        assertFalse(true)
    }
}